### PR TITLE
Mixtral enable

### PR DESCRIPTION
**Pull request by yafshar**
# What does this PR do?

1. Support the mixtral (moe architecture) training
2. Support dynamic shape support for generation, while the static shape has wrong results.

@regisss This PR is originally made by @lkk12014402 https://github.com/huggingface/optimum-habana/pull/668
It is re-based, the mistakes in the docs are fixed and the missing import is being added

### CODE REVIEW

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+            loss_fct = CrossEntropyLoss()
+            shift_logits = shift_logits.view(-1, self.config.vocab_size)
+            shift_labels = shift_labels.view(-1)
+            # Enable model parallelism
+            shift_labels = shift_labels.to(shift_logits.device)
+            loss = loss_fct(shift_logits, shift_labels)
+
+        aux_loss = None
+        if output_router_logits:
+            aux_loss = load_balancing_loss_func(
```
Comment (lkk12014402) [inline optimum/habana/transformers/models/mixtral/modeling_mixtral.py:484]: we should add attention mask. The code is based on transformers==4.37.1 (no attention mask), while the new version 4.37.2 add attention mask.

Comment (regisss) [reply optimum/habana/transformers/models/mixtral/modeling_mixtral.py:484]: We'll add the attention_mask when Transformers v4.38 is released, otherwise we would require users to install Transformers from source to make it work

**Review by regisss — COMMENTED** [review]
Can you run the following from the root of the repo please?
```
pip install -U ruff
make style
```

Also, do you have any insight into the issue with static shapes?

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 2]
```diff
@@ -0,0 +1,569 @@
...
+from transformers.modeling_outputs import MoeCausalLMOutputWithPast, MoeModelOutputWithPast
+from transformers.models.mixtral.modeling_mixtral import MixtralForCausalLM, apply_rotary_pos_emb, repeat_kv
+from transformers.utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+
+
+def gaudi_mixtral_attn_forward(
```
[the remark this answers is not in the document]
Comment (yafshar) [reply optimum/habana/transformers/models/mixtral/modeling_mixtral.py]: @regisss, I fixed the naming as suggested. Do we have to update the mistral naming as well?

[the remark this answers is not in the document]
Comment (regisss) [reply optimum/habana/transformers/models/mixtral/modeling_mixtral.py]: Yes please, good catch!

[the remark this answers is not in the document]
Comment (yafshar) [reply optimum/habana/transformers/models/mixtral/modeling_mixtral.py]: @regisss, I made a PR https://github.com/huggingface/optimum-habana/pull/726 for the Mistral naming correction

**Comment by regisss** [comment]
@yafshar There are some merge conflicts to solve with main, let me know if you need any help

**Comment by regisss** [comment]
@yafshar I just merged #726

**Comment by yafshar** [comment]
@regisss, I resolved the merge conflict with the main branch.

**Comment by regisss** [comment]
@yafshar What is the command you use to test the model? I have not been able to make it work with the text-generation example so far

**Comment by lkk12014402** [comment]
> @yafshar What is the command you use to test the model? I have not been able to make it work with the text-generation example so far

hi, you mean the inference of text-generation?

**Comment by regisss** [comment]
> > @yafshar What is the command you use to test the model? I have not been able to make it work with the text-generation example so far
> 
> hi, you mean the inference of text-generation?

Yep! I managed to have it work on one device in full bf16 with:
```
python run_generation.py --model_name_or_path mistralai/Mixtral-8x7B-v0.1 --use_kv_cache --max_new_tokens 100 --bf16
```
However, it doesn't work on several devices with DeepSpeed (not sure it does on GPU either to be honest). But that's fine to me as this could be done later in another PR if there is no quick fix.

**Comment by yafshar** [comment]
@lkk12014402, @regisss, I am not able to fix the deepspeed. After the optimized transformer injection from deepspeed, the code will break at https://github.com/yafshar/optimum-habana/blob/eae409c3d16508e583fc9dff60dd71342d8160c9/optimum/habana/transformers/models/mixtral/modeling_mixtral.py#L186 If we comment this line and next it will work with no issue. I am currently looking into it.

**Comment by lkk12014402** [comment]
> > > @yafshar What is the command you use to test the model? I have not been able to make it work with the text-generation example so far
> > 
> > 
> > hi, you mean the inference of text-generation?
> 
> Yep! I managed to have it work on one device in full bf16 with:
> 
> ```
> python run_generation.py --model_name_or_path mistralai/Mixtral-8x7B-v0.1 --use_kv_cache --max_new_tokens 100 --bf16
> ```
> 
> However, it doesn't work on several devices with DeepSpeed (not sure it does on GPU either to be honest). But that's fine to me as this could be done later in another PR if there is no quick fix.

hi @regisss
I test generation with this code

```python
import torch
import habana_frameworks.torch.core as htcore
device = torch.device("hpu")

import transformers

from optimum.habana.transformers.modeling_utils import adapt_transformers_to_gaudi
from optimum.habana.transformers.generation.configuration_utils import GaudiGenerationConfig

adapt_transformers_to_gaudi()

# model_name = 'mistralai/Mistral-7B-v0.1'
model_name = 'mistralai/Mixtral-8x7B-v0.1'
model = transformers.AutoModelForCausalLM.from_pretrained(model_name, torch_dtype=torch.bfloat16)
tokenizer = transformers.AutoTokenizer.from_pretrained(model_name)

#model.use_cache = True
model = model.eval().to(device)
# if use_hpu_graphs:
from habana_frameworks.torch.hpu import wrap_in_hpu_graph # pylint: disable=E0401
model = wrap_in_hpu_graph(model)

model.use_cache = True
generation_config = GaudiGenerationConfig(static_shapes=False)

def prepare_inputs(inputs, device):
    return {k:v.to(device=device) for k,v in inputs.items() if torch.is_tensor(v)}

def generate_response(prompt):

    # Format the input using the provided template
    # Tokenize and encode the prompt
    inputs = tokenizer([prompt], return_tensors="pt", add_special_tokens=False)
    inputs = prepare_inputs(inputs, device)

    # Generate a response
    outputs = model.generate(**inputs, generation_config=generation_config, max_length=1000, num_return_sequences=1, do_sample=False, use_cache=True, max_new_tokens=128, eos_token_id=tokenizer.eos_token_id)

    response = tokenizer.decode(outputs[0], skip_special_tokens=False)

    # Extract only the assistant's response
    return response

response = generate_response(system_input)
print(response)
```

**Comment by regisss** [comment]
@yafshar @lkk12014402 I have some changes that make static shapes work, let me see if I can push it directly

**Comment by regisss** [comment]
@lkk12014402 Can you try now with static shapes?

**Comment by lkk12014402** [comment]
> @lkk12014402 Can you try now with static shapes?

hi regiss, I understand your changes. But when we remove "continue", all experts are computed, so the inference performance maybe slower.

I try your revised code, but get some error

- with your command:
python run_generation.py --model_name_or_path mistralai/Mixtral-8x7B-v0.1 --use_kv_cache --max_new_tokens 100 --bf16

image

- with my inference code:
image

Actually, at the previous version code, we can also remove the "continue" control flow, it works well for static shape.

**Comment by regisss** [comment]
@lkk12014402 Weird, I don't get these errors :thinking: 

I pushed a new commit that keeps the `continue` call but bypasses it if we are using HPU graphs.

**Comment by yafshar** [comment]
> ```python
> system_input
> ```
@regisss, The latest commit is working fine with hpu_graphs and without it. I am still working on the deepspeed issue!

**Comment by regisss** [comment]
Habana also opened a PR for this model: #739 
It basically does the same as this one + a few other optimizations. It may be better to contribute there now!

**Comment by yafshar** [comment]
I am closing the PR. Since https://github.com/huggingface/optimum-habana/pull/739 does the same and other optimizations
